iree/compiler/Dialect/Flow/IR/FlowOpUtils.h

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Operation.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+// Clones an operation with new result types.
+// The original operation will be erased and a new operation constructed
+// in its place.
+Operation *cloneWithNewResultTypes(Operation *op,
+                                   llvm::SmallVectorImpl<Type> &newResultTypes);
+
+// Utility class to optimize a "closure" op, which maintains a variadic
+// list of operands corresponding to entry block arguments.
+class ClosureOpDce {
+ public:
+  ClosureOpDce(Operation *closureOp, Block &entryBlock,
+               unsigned variadicOffset);
+
+  bool needsOptimization() { return needsOperandElision || needsResultElision; }
+
+  // Whether the operation needs to be replaced.
+  bool needsNewOperation() { return needsResultElision; }
+
+  // Performs the optimization. If the optional eraseOriginal=false and
+  // needsNewOperation(), then the original will not be erased, leaving that
```

Comment:
If you were able to take a PatternRewriter then you could really simplify this, as then the caller wouldn't need to do the needsNewOperation/branch dance. If you still need to support non-PatternRewriter you could just add a static optimizeClosureOp() method matchAndRewrite would call and at least make that simpler. This looks like a super handy tool and making it easier to excise/make more central in the future would be nice :)

Reply:
Per note above, agreed and will get this fixed correctly upstream.